Add WhisperC++ engine to speech-to-text-impl

majosch opened the pull request:
> ~~Caveat: This PR is not ready to merge and needs to be debugged - command execution still fails - see this [post](https://groups.google.com/a/opencast.org/g/users/c/AzJYC0QhOz0/m/Pk3QujakAQAJ) in Opencast users group.~~ As suggested by @mliradelc I opened this PR for easier reviewing.

This PR adds a WhisperC++ engine to the speech-to-text-module.

[WhisperC++](https://github.com/ggerganov/whisper.cpp) is a C/C++ implementation of OpenAI's Whisper automatic speech recognition (ASR) model.

Many thanks to @mliradelc and @geichelberger for any help so far!

Commit:
fix whispercpp engine bug

sulyoklev committed:
in WhisperCppEngine: fix JSON not correctly reading language after subtitle generation; add exception handling that is thrown if files formats other than wav are used as input; add language-detection as default if no language is specified

KatrinIhler commented:
Is this ready yet? If not, this should probably be a draft.

mliradelc commented:
It was ready, but I had problems testing it, at the end Ctranslate2 PR arrived and overwrited some things regarding this. The autor needs to rebase.

Commit:
add config options beam size and max segment length to whispercpp

Commit:
optional config parameters

Commit:
update docs and use track instead attachment in example

majosch commented:
Thanks to @sulyoklev for rebasing the PR.

@KatrinIhler PR should be ready for review now.

Works fine on our test installation (15-SNAPSHOT).

If someone else wants to test:
- static binary: [whispercpp.zip](https://github.com/opencast/opencast/files/13112158/whispercpp.zip)
- tiny model: https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-tiny.bin

Commit:
add WhisperC++ engine to speech-to-text-impl

Commit:
update docs after PR 5352

mliradelc commented:
Hi, I started today to review this PR, but the dependencies are playing against me 😅 . At least I've got anaconda started to compile the coreml models but another missing dependency started to make problems 😓 .

lkiesow committed:
Fix broken documentation links

This patch fixes a few broken links in the documentation caused by some
:x

lkiesow commented:
Hi @majosch, I did a few tests and did a rebase of your code. You will find the rebase at https://github.com/lkiesow/opencast/tree/whisper-cpp
Can you update this pull request? I'll then do a proper review and we can try getting this in.

lkiesow committed:
Consume output of whisper

This patch ensures that the output of the whisper command is actually
being consumed. Otherwise, the process may get stuck while processing
larger files.

lkiesow commented:
Thanks for updating this, @majosch. @mwuttke tested your patch from yesterday and found some issues. If you throw in long recordings, the whole process can hang indefinitely. I pushed a commit to https://github.com/lkiesow/opencast/tree/whisper-cpp (specifically https://github.com/lkiesow/opencast/commit/7be1348fe8cfdaeeb7f1dd74a7f6d6629bbe34d5) which should fix the issue.

lkiesow commented:
Looking at the failed tests, you are also missing updates to your docs which I already included in my rebase at https://github.com/lkiesow/opencast/commit/5ea4aadef1ad538c56415ade87beefa7dbc62b1c. Not aure how you updated your branch, but you will need something similar.

majosch commented:
@lkiesow Thank you for looking into this PR and for the fix. Also thanks to @mwuttke for testing.

I just updated r/15.x on our Opencast fork and did a `git rebase r/15.x" and fixed the merge conflicts - so, if I made no mistake, links should hopefulle already be correct.

Would you prefer updating my branch with yours, Lars?

majosch commented:
@lkiesow I just push your branch.

lkiesow commented on docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md:
```diff
@@ -0,0 +1,73 @@
+
+1. Install whispercpp binary and language models following the instruction on
+[whisper.cpp](https://github.com/ggerganov/whisper.cpp) repository or build and use your own RPMs for RHEL-based
+distributions from [whispercpp-rpmbuild](https://github.com/elearning-univie/whispercpp-rpmbuild).
```
```suggestion
[whisper.cpp](https://github.com/ggerganov/whisper.cpp) repository or install the `whisper.cpp` package from the Opencast repository.
```

lkiesow commented on docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md:
```diff
@@ -0,0 +1,73 @@
+  - id: encode
+    description: "Extract audio for processing with whispercpp"
+    configurations:
+      - source-flavor: "*/themed"
```
Most of the time, you probably want to operate on `*/source`:
```suggestion
      - source-flavor: "*/source"
```

lkiesow commented on docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md:
```diff
@@ -0,0 +1,73 @@
+      - target-tags: >-
+          archive,
+          engage-download
+      # - translate: true
```
```suggestion
```

lkiesow commented on docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md:
```diff
@@ -0,0 +1,73 @@
+# Audio-only encoding format used for whispercpp
+profile.audio-whispercpp.name = whispercpp wav
+profile.audio-whispercpp.input = audio
+profile.audio-whispercpp.output = audio
```
iirc, this won't work with a video as input.
I think this should be:
```suggestion
profile.audio-whispercpp.input = stream
profile.audio-whispercpp.output = stream
```

lkiesow commented on docs/guides/admin/docs/configuration/transcription.configuration/whispercpp.md:
```diff
@@ -0,0 +1,73 @@
+  binary/library is built with appropriate options/modules.
+- Language models are based on Whisper and are transformed to a custom ggml format. You can download already converted
+  models or build and tweak your own models. For more informations visit
+  [models/README.md](https://github.com/ggerganov/whisper.cpp/blob/master/models/README.md).
```
Note to self: Add a point about packages and the packaged version of the download script

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/speechtotext-woh.md:
```diff
@@ -6,23 +6,25 @@ ID: `speechtotext`
-The speech to text operation can be used to generate subtitles for Videos or Audio files. Currently, there are two STT
-engines available, [Whisper](../configuration/transcription.configuration/whisper.md) and 
-[Vosk](../configuration/transcription.configuration/vosk.md). The subtitles file format ist WebVTT.
+The speech to text operation can be used to generate subtitles for Videos or Audio files. Currently, there are three STT
```
```suggestion
The speech to text operation can be used to generate subtitles from video or audio files. Currently, there are three STT
```

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+  }
+
+  //TODO: Add method for language detection
+
```
Do it or remove this ;-)
```suggestion
```

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+   * @see org.opencastproject.speechtotext.api.SpeechToTextEngine#generateSubtitlesFile(File, File, String, Boolean)
+   */
+
+  @Override
```
```suggestion
   */
  @Override
```

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+      File preparedOutputFile, String language, Boolean translate)
+          throws SpeechToTextEngineException {
+
+    if (!FilenameUtils.isExtension(mediaFile.getPath(), "wav")) {
```
Not a big deal, but maybe to be a bit more flexible:
```suggestion
    if (!FilenameUtils.isExtension(mediaFile.getPath(), "wav")) {
    if (!mediaFile.getPath().toLowerCase().endsWith(".wav")) {
```

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+      throw new SpeechToTextEngineException("WhisperC++ currently doesn't support any media extension other than wav");
+    }
+
+    String[] baseCommands = { whispercppExecutable,
+        mediaFile.getAbsolutePath(),
+        "--model", whispercppModel,
+        "-ovtt",
+        "-oj",
+        "--output-file", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")};
+
+    List<String> command = new ArrayList<>(Arrays.asList(baseCommands));
```
Instead of creating an array, converting that to a list which is then used to initialize an `ArrayList` below, you could directly work with a list. Something like:
```suggestion
    List<String> baseCommand = List.of(whispercppExecutable,
        mediaFile.getAbsolutePath(),
        "--model", whispercppModel,
        "-ovtt",
        "-oj",
        "--output-file", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", ""));

    List<String> command = new ArrayList<>(baseCommand);
```

lkiesow replied:
Or even:
```suggestion
    var command = new ArrayList<String>(List.of(
        whispercppExecutable,
        mediaFile.getAbsolutePath(),
        "--model", whispercppModel,
        "-ovtt",
        "-oj",
        "--output-file", preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")));
```

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+    }
+    if (whispercppWordThreshold.isSome()) {
+      command.add("-wt");
+      command.add(String.format("%f", whispercppWordThreshold.get()));
```
This depends on the system's language settings and may break.
If the system language is English, this will result in something like `1.1`:

```
❯ echo $LANG
en_US.UTF-8

❯ jshell    
jshell> String.format("%f", 1.1)
$1 ==> "1.100000"
```

But switching to German, for example, this will be `1,1` which I assume will cause a problem?
```
❯ export LANG=de_DE.UTF-8

❯ jshell                 
jshell> String.format("%f", 1.1)
$1 ==> "1,100000"
```

This is used at more places down below.

You are also converting the configuration values to numbers, only to convert them back here. You could just never convert them.

sulyoklev replied:
You're right. Converting the configuration values to numbers, although somewhat unnecessary, is useful to prevent unwanted behavior when incompatible config values are passed to whispercpp.  This way it is ensured that whispercpp will behave as expected. I will fix the format issue.

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+      logger.info("Using language {} from workflows", language);
+      command.add("--language");
+      command.add(language);
+    } else {
+      logger.debug("Auto-detecting language");
+      command.add("--language");
+      command.add("auto");
```
I think this is superfluous since auto-detection is the default anyway. There is no need to pass on this parameter, is there?

sulyoklev replied:
Actually no, the default in whispercpp is English, so it's needed for auto detection

gregorydlogan replied:
So my setup is defaulting to 
` --language eng`
which spits out 
```
whisper_lang_id: unknown language 'eng'
error: unknown language 'eng'
```

By default, if I did not set the language in the workflow I would get `eng`, rather than `english`, or `auto`.

sulyoklev replied:
It seems the workflow defaults to using `eng` if the language code isn't given, eventhough the speechtotext-woh documentation says that language codes are the two latter codes defined in ISO 639-1. Since whispercpp only accepts either the two latter codes (e.g. `en`) or the full lowercase names (e.g. `english`) it cannot recognize `eng` and thus throws an error.

lkiesow commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:
```diff
@@ -0,0 +1,417 @@
+    if (subtitleLanguage.isBlank()) {
+      JSONParser jsonParser = new JSONParser();
+      try {
+        FileReader reader = new FileReader(preparedOutputFile.getAbsolutePath().replaceFirst("[.][^.]+$", "")
+            + ".json");
```
I'm writing this down here since this operated on the JSON file, although this is a more general problem. This code will cause Whisper to generate a JSON file which is never cleaned up. The changes made necessary by pull request #5472 will also fix this and you wouldn't need to clean up every file, but waiting for that PR also means some minor changes to this methods interface.

sulyoklev replied:
I added the cleanup, so this PR can be merged anytime before or after #5472

lkiesow reviewed (changes requested):
Sorry for the delay. I finally found time to do a proper review of this.
This generally looks good. I just found a few minor issues.

Depending on pull request #5472 being merged before/after this one of the pull requests needs to fix some conflicts as well.

Some of the comments are just suggestions. Feel free to argue against any of these changes ;-)

Commit:
cleanup json file generated by whispercpp; fix format locale issue of float config parameters in whispercpp

majosch committed:
update whispercpp docs

majosch committed:
fix preposition and use lowercase in speech-to-text-woh docs

gregorydlogan commented:
Trying to take a look at this, but I'm seeing compilation issues:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project opencast-speech-to-text-impl: Compilation failure: Compilation failure: 
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[62,8] org.opencastproject.speechtotext.impl.engine.WhisperCppEngine is not abstract and does not override abstract method generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.api.SpeechToTextEngine
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[260,30] generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.impl.engine.WhisperCppEngine cannot implement generateSubtitlesFile(java.io.File,java.io.File,java.lang.String,java.lang.Boolean) in org.opencastproject.speechtotext.api.SpeechToTextEngine
[ERROR]   return type java.util.Map<java.lang.String,java.lang.Object> is not compatible with org.opencastproject.speechtotext.api.SpeechToTextEngine.Result
[ERROR] /home/greg/opencast/upstream/modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/engine/WhisperCppEngine.java:[259,3] method does not override or implement a method from a supertype

```

Commit:
Adapt engine to api changes in #5472

sulyoklev commented:
Should compile now, I've implemented the changes in #5472.

gregorydlogan commented on etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg:
```diff
@@ -3,5 +3,5 @@
-# Available engines: vosk, whisper | Default: vosk
-SpeechToTextEngine.target=(enginetype=whisper)
+# Available engines: vosk, whisper, whispercpp | Default: vosk
+#SpeechToTextEngine.target=(enginetype=vosk)
```
So, this config doesn't actually *do* anything, since `STTSImpl` does not parse it, or change its hardcoded (vosk) binding.

What this does do is cause `STTSImpl` to hang waiting for its bindings to finish :)

sulyoklev replied:
You're right, I pushed a fix

gregorydlogan reviewed (commented):
In testing this works well, aside from my comments.

Commit:
fix STTS config to bind engine type

gregorydlogan commented on etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg:
```diff
@@ -4,4 +4,4 @@
 # Select STT engine.
 # Available engines: vosk, whisper, whispercpp | Default: vosk
-#SpeechToTextEngine.target=(enginetype=vosk)
+SpeechToTextEngine.target=(enginetype=whisper)
```
A few of issues here:
- You meant `whispercpp` here I'm assuming?
- This changes the default config, and thus can't go into `r/15.x`
- This config change still does not work how you want it to.

`SpeechToTextServiceImpl` does not read this config and thus can't actually change its binding filter.  You want to add logic to its `activate` method, where both the `@activate` and `@modified` annotations live.  And probably rip out/replace `setSpeechToTextEngine` in the same class, since it's where the hardcoded binding lives.

majosch replied:
This PR does NOT intend to change the default SpeechToText-Engine - the changed line just reflects the changes from 14 to 15.

gregorydlogan replied:
I think it's probably https://github.com/opencast/opencast/pull/4973/commits/014827467f70d8a650cdefc6063dc2327fea8d6f that's causing the problem.  I saw that listed in this PR, and assumed it was an attempt to resolve the binding issue from the commit log entry.  Your commit seems to duplicate 0e2c9ba3086a2d6093a5ed3f28709cfb149de1e3 (which probably shouldn't have touche this file, but I digress), which is already in r/15.x

mwuttke commented:
I have successfully tested the current status of this patch with the Opencast Version 15.2 together with the language packs that @gregorydlogan has built.

Would it be possible to merge the patch quickly so that it is included in the next release?

Or what else speaks against a merge?

gregorydlogan commented:
@mwuttke Did you have to modify the patch at all?  Notably, `etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg`'s target config does not seem to work for me.

Otherwise I think this is reasonable, and I suspect that the config item I'm talking about it just straight up *broken* for everything so I don't see that preventing this from being merged.  @lkiesow short notice, but any objections?

mwuttke commented:
I have only changed the enginetype in `/etc/org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.cfg` to whispercpp. That was all. After that, the transcription ran without errors.

gregorydlogan commented:
Retested, and the engine config does actually work.

I did see a failure when the language was not set for my workflow but that easily cleared up, a default can be set, and this engine isn't even active by default anyway.  Merging.
